File: libs/monoboard/src/main/kotlin/mono/graphics/board/Pixel.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.graphics.board

import mono.common.Characters.TRANSPARENT_CHAR
import mono.common.Characters.isHalfTransparent
import mono.common.Characters.isTransparent

/**
 * A class represents each item on the board matrix.
 */
class Pixel(
    visualChar: Char = TRANSPARENT_CHAR,
    directionChar: Char = TRANSPARENT_CHAR,
    highlight: Highlight = Highlight.NO
) {
    var visualChar: Char = visualChar
        private set

    var directionChar: Char = directionChar
        private set

    var highlight: Highlight = highlight
        private set

    val isTransparent: Boolean
        get() = visualChar.isTransparent || visualChar.isHalfTransparent

    fun set(visualChar: Char, highlight: Highlight) {
        this.visualChar = visualChar
        // TODO: Correct this with a direction char
        this.directionChar = visualChar
        this.highlight = highlight
    }

    fun reset() {
        visualChar = TRANSPARENT_CHAR
        directionChar = TRANSPARENT_CHAR
        highlight = Highlight.NO
    }

    override fun toString(): String = if (isTransparent) " " else visualChar.toString()

    override fun equals(other: Any?): Boolean {
        return other is Pixel && visualChar == other.visualChar && highlight == other.highlight
    }

    override fun hashCode(): Int {
        var result = visualChar.hashCode()
        result = 31 * result + highlight.hashCode()
        return result
    }

    companion object {
        val TRANSPARENT_PIXEL = Pixel()
    }
}

/**
 * An enum to describe all kinds of highlight
 */
enum class Highlight {
    NO,
    SELECTED,
    TEXT_EDITING
}

File: monosketch-svelte/src/lib/mono/monobitmap/board/pixel.ts
import type { Char } from '$libs/char';
import { isHalfTransparentChar, isTransparentChar, TRANSPARENT_CHAR } from '$mono/common/character';
import type { Comparable } from '$libs/comparable';

/**
 * An interface for the pixel of the Mono Bitmap.
 */
export class Pixel implements Comparable {
    static TRANSPARENT = new Pixel();

    private visualCharInner: Char = TRANSPARENT_CHAR;
    private directionCharInner: Char = TRANSPARENT_CHAR;
    private highlightInner: HighlightType = HighlightType.NO;

    get isTransparent(): boolean {
        const char = this.visualCharInner;
        return isTransparentChar(char) || isHalfTransparentChar(char);
    }

    get visualChar(): Char {
        return this.visualCharInner;
    }

    get directionChar(): Char {
        return this.directionCharInner;
    }

    get highlight(): HighlightType {
        return this.highlightInner;
    }

    set = (visualChar: Char, directionChar: Char, highlight: HighlightType) => {
        this.visualCharInner = visualChar;
        this.directionCharInner = directionChar;
        this.highlightInner = highlight;
    };

    reset = () => {
        this.visualCharInner = TRANSPARENT_CHAR;
        this.directionCharInner = TRANSPARENT_CHAR;
        this.highlightInner = HighlightType.NO;
    };

    toString = (): string => (this.isTransparent ? ' ' : this.visualChar);

    equals(other: unknown): boolean {
        if (!(other instanceof Pixel)) {
            return false;
        }
        const otherPixel = other as Pixel;
        return (
            this.visualCharInner === otherPixel.visualCharInner &&
            this.highlightInner === otherPixel.highlightInner
        );
    }
}

export enum HighlightType {
    NO,
    SELECTED,
    TEXT_EDITING,
    LINE_CONNECT_FOCUSING,
}
